[Mapper] Add `ignore_missing` option to `timestamp`

dadoonet opened the pull request:
Related to #9049.

By default, the default value for `timestamp` is `now` which means the date the document was processed by the indexing chain.

You can now reject documents which not provide a `timestamp` value by setting `ignore_missing` to false (default to `true`):

```js
{
    "tweet" : {
        "_timestamp" : {
            "enabled" : true,
            "ignore_missing" : false
        }
    }
}
```

When you update the cluster to 1.5 or master, this index created with 1.4 we automatically migrate an index created with 1.4 to the 1.5 syntax.

Let say you have defined this in elasticsearch 1.4.x:

```js
DELETE test
PUT test
{
  "settings": {
    "number_of_shards": 1,
    "number_of_replicas": 0
  }
}
PUT test/type/_mapping
{
  "type" : {
      "_timestamp" : {
          "enabled" : true,
          "default" : null
      }
  }
}
```

After migration, the mapping become:

```js
{
   "test": {
      "mappings": {
         "type": {
            "_timestamp": {
               "enabled": true,
               "store": false,
               "ignore_missing": false
            },
            "properties": {}
         }
      }
   }
}
```

Closes #8882.

rjernst commented:
LGTM.

dadoonet commented:
@rjernst based on what we saw in #9233, I think we should at startup try to read an old 1.4 mapping and if default was explicitely set to null, add on the fly ignore_missing=false. I need to run a test but I believe that reading a 1.4 mapping with default = null might fail at startup in 1.5. WDYT?

es-github-notifications commented:
It is unclear to me what the null behavior is supposed to be for the
settings there?  Is null a valid value for any of them? I would rather give
an error if the value is invalid for 1.5+, but allow reading null from old
mappings for backcompat (and skip it, although with a warning).

On Sat, Jan 10, 2015 at 9:03 PM, David Pilato <notifications@github.com>
wrote:

> @rjernst <https://github.com/rjernst> based on what we saw in #9233
> <https://github.com/elasticsearch/elasticsearch/pull/9233>, I think we
> should at startup try to read an old 1.4 mapping and if default was
> explicitely set to null, add on the fly ignore_missing=false. I need to run
> a test but I believe that reading a 1.4 mapping with default = null might
> fail at startup in 1.5. WDYT?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elasticsearch/elasticsearch/pull/9104#issuecomment-69483850>
> .
>

dadoonet commented:
@rjernst I rebased my PR on master and added a new commit as what I thought about actually happened. It was not possible to read an index created in elasticsearch 1.4 with elasticsearch 1.5 or above with this PR. The commit fixed that.
Could you review it again please - at least the last commit as you already reviewed the first one? Thank you!

dadoonet commented:
BTW I think I should perhaps add in documentation that with 1.5, when reading an old mapping created with 1.4, we actually modify it behind the scene. WDYT?

rjernst commented:
The added change LGTM.  In regards to 1.5 docs, I think being explicit about what we are doing would be good.
